### ISSUE TITLE
Erroneous join in Update queries

### ISSUE THREAD

**Issue by yegorpetrov** [zzzprojects/EntityFramework-Plus#69]
In one of my projects BatchUpdate generates erroneous queries like this one:

	UPDATE A 
	SET A.[ItemJustification] = @zzz_BatchUpdate_0
	FROM [dbo].[Record] AS A
	INNER JOIN ( SELECT 
		[Extent1].[Id] AS [Id], 
		[Extent2].[Id] AS [Id1]
		-- many irrelevant fields from Extent2 are omitted
		FROM  [dbo].[Task] AS [Extent1]
		INNER JOIN [dbo].[Record] AS [Extent2] ON [Extent1].[RecordId] = [Extent2].[Id]
		WHERE [Extent1].[Id] IN (3002)
			   ) AS B ON A.[Id] = B.[Id]

At the last line it should be `A.[Id] = B.[Id1]`.

Unfortunately, I was unable to reproduce this from scratch in a console app because the sub-query ends up not having Extent1.Id/Extent2.Id1 aliases:

	UPDATE A
	SET A.[FieldB] = @zzz_BatchUpdate_0
	FROM [dbo].[EntityBs] AS A
	INNER JOIN ( SELECT
		[Extent2].[Id] AS [Id],
		[Extent2].[FieldB] AS [FieldB]
		FROM  [dbo].[EntityAs] AS [Extent1]
		INNER JOIN [dbo].[EntityBs] AS [Extent2] ON [Extent1].[EntityBId] = [Extent2].[Id]
		WHERE [Extent1].[Id] IN (1)
			   ) AS B ON A.[Id] = B.[Id]

Here's the minimal code to show you what's going on. As I said, the bug doesn't show up here.

    public class EntityA
    {
        public int Id { get; set; }
        public string FieldA { get; set; }
        public int EntityBId { get; set; }
        public virtual EntityB EntityB { get; set; }
    }

    public class EntityB
    {
        public int Id { get; set; }
        public string FieldB { get; set; }
        public virtual ICollection<EntityA> EntitiesA { get; set; } = new HashSet<EntityA>();
    }

    public class EDbContext : DbContext
    {
        public EDbContext() : base("DataContext") { }

        public DbSet<EntityA> EntityA { get; set; }
        public DbSet<EntityB> EntityB { get; set; }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            modelBuilder.Entity<EntityA>().HasKey(e => e.Id);
            modelBuilder.Entity<EntityB>().HasKey(e => e.Id);
            modelBuilder.Entity<EntityB>().HasMany(e => e.EntitiesA).WithRequired(e => e.EntityB);
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx1 = new EDbContext())
            {
                if (!ctx1.Set<EntityA>().Any()) // seed
                {
                    ctx1.Set<EntityB>().Add(new EntityB
                    {
                        FieldB = "Id 1 dummy"
                    });
                    ctx1.SaveChanges();
                    ctx1.Set<EntityA>().Add(new EntityA
                    {
                        // Id 1
                        FieldA = "AAA",
                        EntityB = new EntityB
                        {
                            // Id 2
                            FieldB = "___"
                        }
                    });
                    ctx1.SaveChanges();
                }
            }

            using (var ctx2 = new EDbContext())
            {
                var id = new[] { ctx2.Set<EntityA>().FirstOrDefault().Id };
                var iq = ctx2.Set<EntityA>().Where(a => id.Contains(a.Id));
                iq.Select(a => a.EntityB).Update(b => new EntityB { FieldB = "BBB" }, bu => bu.Executing = Exec);
            }

            using (var ctx3 = new EDbContext())
            {
                Debug.Assert(ctx3.Set<EntityA>().FirstOrDefault().EntityB.FieldB == "BBB");
            }
        }

        private static void Exec(DbCommand cmd)
        {
            Console.WriteLine(cmd.CommandText);
        }

**Comment by zzzprojects** [comment]
Hello @yegorpetrov ,

I tried to replicate the bug unsuccessfully for a few days. The [Extent1].[Id] AS [Id] never show up!

Can you post or send us (info@zzzprojects.com) the full LINQ Query you used without omitting anything? I'm sure that will help us to find something we forget to do to replicate the issue.

Best Regards,

Jonathan

**Comment by zzzprojects** [comment]
Hello @yegorpetrov ,

We still have not received anything from your part.

Best Regards,

Jonathan

**Comment by yegorpetrov** [comment]
Yeah, sorry about the delay. The problem is highlighted on this screenshot:

![batchupdateproblem](https://cloud.githubusercontent.com/assets/10397442/20139738/6c71d9c4-a6a9-11e6-8582-0b4f6710c411.png)

This thing is proprietary, so I'm not sure I'm free to share it in the open. But let me know if the screenshot is of any help, and may be we'll figure out what I can send to you privately.

**Comment by zzzprojects** [comment]
Hello @yegorpetrov ,

What I would like is the screenshot of the LINQ query if you can. That would maybe give me a hint to how to build the LINQ query/entities to reproduce it and generate this SQL.

Like I said, you can send this information to us in private: info@zzzprojects.com

Best Regards,

Jonathan

**Comment by yegorpetrov** [comment]
Well, it's already shown on that screenshot. `_taskRepo.Table is essentially a DbSet<Task>` from the context, and you can see the rest. Not sure what other LINQ I should send to you.

**Comment by zzzprojects** [comment]
Oh sorry, you are right. I checked it too fast.
